File: Cpp/STL/238.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> productExceptSelf(vector<int> &nums)
{
    int prod = 1;
    for (int i = 0; i < nums.size(); i++)
    {
        prod *= nums.at(i);
    }
    vector<int> ans;
    for (int i = 0; i < nums.size(); i++)
    {
        if (nums.at(i) != 0)
            ans.push_back(prod / nums.at(i));
        else
            ans.push_back(prod);
    }
    return ans;
}

int main()
{
    vector<int> nums = {1, 2, 3, 4};
    vector<int> ans = productExceptSelf(nums);

    for (int i : ans)
        cout << i << " ";
}

File: Cpp/STL/medium/238.py
from typing import List

def productExceptSelf(nums: List[int]) -> List[int]:
    left = [1]
    right = [1]*len(nums)
    
    for i in range(1,len(nums)):
        left.append(left[i-1]*nums[i-1])
    
    for i in range(len(nums)-2,-1,-1):
        right[i] = right[i+1]*nums[i+1]
    # print(left, right)

    ans = []
    for i in range(len(nums)):
        ans.append(left[i]*right[i])
    return ans
        

nums = [1,2,3,4]
print(productExceptSelf(nums))